File: sentinel_gateway/server/models/car_fleet/device/cfDevice.model.js
import mongoose from 'mongoose';

const cfDeviceSchema = new mongoose.Schema({
  name: {
    type: String,
    reguired: true,
    trim: true,
    unique: true,
    set: function(name) {
      this._name = this.name;
      return name;
    },
  },
  device_id: {
    type: String,
    reguired: true,
    trim: true,
    unique: true,
    uppercase: true,
    set: function(device_id) {
      this._device_id = this.device_id;
      return device_id;
    },
  },
  mac_address: {
    type: String,
    reguired: true,
    trim: true,
    unique: true,
    uppercase: true,
    mixLength: 12,
    maxLength: 17,
    set: function(mac_address) {
      this._mac_address = this.mac_address;
      return mac_address;
    },
  },
  car: { // TODO - hooks to handle other side
    type: mongoose.Types.ObjectId,
    ref: 'cf_car',
    unique: true,
    sparse: true,
    set: function(car) {
      this._car = this.car;
      return car;
    },
  },
  description: {
    type: String,
    trim: true,
    set: function(description) {
      this._description = this.description;
      return description;
    },
  },
}, {
  timestamps: true,
  toObject: { virtuals: true },
  toJSON: { virtuals: true},
});

export default mongoose.model('cf_device', cfDeviceSchema);

File: sentinel_client/lib/models/cfDevice.model.dart
// import 'package:sentinel_client/models/cfCar.model.dart';

class CfDevice {
  final String id;
  /*final*/ String name;
  /*final*/ String deviceId;
  /*final*/ String macAddress;
  // final CfCar car;
  /*final*/ String? carId; // TODO
  /*final*/ String? description;

  CfDevice(
    this.id,
    this.name,
    this.deviceId,
    this.macAddress,
    this.carId,
    this.description,
  );

  CfDevice.fromJson(Map<String, dynamic> json)
      : id = json['_id'],
        name = json['name'],
        deviceId = json['device_id'],
        macAddress = json['mac_address'],
        carId = json['car'],
        description = json['description'];
}
